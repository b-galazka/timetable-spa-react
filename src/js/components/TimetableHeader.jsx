import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import propTypes from 'prop-types';

import urlTranslations from '../../json/urlTranslations';

import '../../scss/timetableHeader.scss';

function mapStateToProps(state) {

    const { teachers } = state;

    return { teachers };
}

class TimetableHeader extends Component {

    render() {

        return (
            <header className="timetable__header">
                <h1 className="timetable__title">
                    {this.renderTitle()}
                </h1>
            </header>
        );
    }

    renderTitle() {

        const { type: urlParam } = this.props.match.params;
        const timetableObjectType = urlTranslations[urlParam];

        if (timetableObjectType === 'teacher') {

            return this.renderTeacherTitle();
        }

        return this.getSlug();
    }

    renderTeacherTitle() {

        const teacherSlug = this.getSlug();
        const { teachers } = this.props;
        const currentTeacher = teachers.find(teacher => teacher.slug === teacherSlug);

        return currentTeacher.name || currentTeacher.slug;
    }

    getSlug() {

        return decodeURIComponent(this.props.match.params.slug);
    }
}

TimetableHeader.propTypes = {

    // router
    match: propTypes.shape({
        params: propTypes.shape({
            slug: propTypes.string.isRequired,
            type: propTypes.string.isRequired
        }).isRequired
    }).isRequired,

    // redux
    teachers: propTypes.arrayOf(
        propTypes.shape({
            slug: propTypes.string.isRequired,
            name: propTypes.string
        }).isRequired
    ).isRequired
};

export default withRouter(connect(mapStateToProps)(TimetableHeader));