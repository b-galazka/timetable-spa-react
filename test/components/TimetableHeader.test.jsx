import React from 'react';
import renderer from 'react-test-renderer';

import TimetableHeader from '../../src/js/components/TimetableHeader';

jest.mock(
    '../../src/json/urlTranslations',
    () => ({
        teacher: 'teacher',
        class: 'class',
        classroom: 'classroom'
    })
);

describe('TimetableHeader component', () => {

    it('should render class\'s title', () => {

        const tree = renderer.create(
            <TimetableHeader.WrappedComponent

                match={{
                    params: { slug: '%233ET', type: 'class' }
                }}

                teachers={[
                    { slug: 'XYZ', name: null }
                ]}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render classroom\'s title', () => {

        const tree = renderer.create(
            <TimetableHeader.WrappedComponent

                match={{
                    params: { slug: '%23200', type: 'classroom' }
                }}

                teachers={[
                    { slug: 'XYZ', name: null }
                ]}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render teacher\'s title (slug)', () => {

        const tree = renderer.create(
            <TimetableHeader.WrappedComponent

                match={{
                    params: { slug: '%23XYZ', type: 'teacher' }
                }}

                teachers={[
                    { slug: '#XYZ', name: null }
                ]}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render teacher\'s title (name)', () => {

        const tree = renderer.create(
            <TimetableHeader.WrappedComponent

                match={{
                    params: { slug: 'XYZ', type: 'teacher' }
                }}

                teachers={[
                    { slug: 'XYZ', name: 'XYZ name' }
                ]}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});