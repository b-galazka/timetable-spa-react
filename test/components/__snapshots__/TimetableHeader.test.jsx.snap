// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimetableHeader component should render class's title 1`] = `
<header
  className="timetable__header"
>
  <h1
    className="timetable__title"
  >
    #3ET
  </h1>
</header>
`;

exports[`TimetableHeader component should render classroom's title 1`] = `
<header
  className="timetable__header"
>
  <h1
    className="timetable__title"
  >
    #200
  </h1>
</header>
`;

exports[`TimetableHeader component should render teacher's title (name) 1`] = `
<header
  className="timetable__header"
>
  <h1
    className="timetable__title"
  >
    XYZ name
  </h1>
</header>
`;

exports[`TimetableHeader component should render teacher's title (slug) 1`] = `
<header
  className="timetable__header"
>
  <h1
    className="timetable__title"
  >
    #XYZ
  </h1>
</header>
`;
